torchtune/data/_chat_formats.py

```diff
@@ -0,0 +1,215 @@
+            content = ""
+            if message.role == "system":
+                content = cls.system.format(content=message.content)
+                system_message = content
+                # Incorporate the system message in the user message - LLaMA2 only
+                # looks for the <<SYS>> tags and not the explicit role so this will
+                # be treated the same as an actual system message. We do this because
+                # of the nesting of the system prompt in the user message.
+                continue
+            elif message.role == "user":
+                content = cls.user.format(
+                    system_message=system_message, content=message.content
+                )
+            elif message.role == "assistant":
+                # No special formatting needed for assistant message
+                content = message.content
+            assert content != ""
+            formatted_dialogue.append(
+                Message(role=message.role, content=content, masked=message.masked),
+            )
+        return formatted_dialogue
+
+
+class MistralChatFormat(ChatFormat):
+    """
+    Formats according to Mistral's instruct model:
+    https://docs.mistral.ai/models/
+
+    It is identical to `Llama2ChatFormat`, except it does not support system
+    prompts.
+
+    Example:
+        "[INST] I am going to Paris, what should I see? [/INST] Paris, the capital
+        of France, is known for its stunning architecture..."
+
+    """
+
+    B_INST, E_INST = "[INST]", "[/INST]"
+    system = ""
+    user = f"{B_INST} {{content}} {E_INST} "
+    assistant = ""
+
+    @classmethod
+    def format(
+        cls,
+        sample: List[Message],
+    ) -> List[Message]:
+        """
+        Format user and system messages with appropriate tags.
+
+        Args:
+            sample (List[Message]): a single conversation, structured as a list
+                of `Message` objects
+
+        Returns:
+            The formatted list of messages
+
+        Raises:
+            ValueError: If system prompts are provided
+        """
+        formatted_dialogue = []
+        for message in sample:
+            content = ""
+            if message.role == "system":
+                raise ValueError(
+                    "System prompts are not supported in MistralChatFormat"
+                )
+            elif message.role == "user":
+                content = cls.user.format(
+                    content=message.content,
+                )
+            elif message.role == "assistant":
+                # No special formatting needed for assistant message
+                content = message.content
+            assert content != ""
+            formatted_dialogue.append(
+                Message(role=message.role, content=content, masked=message.masked),
+            )
+        return formatted_dialogue
+
+
+class ChatMLFormat(ChatFormat):
+    """
+    OpenAI's Chat Markup Language used by their chat models:
```

Comment:
As far as I know, OpenAI models support only tiktoken. Are we adding support for that in this PR or in a follow up? Or not yet?

Reply:
Let's leave it for a follow-up if you're good with that

Reply:
I believe they use this format with TikToken: https://community.openai.com/t/how-does-chatml-do-the-exact-formatting/80751

The main motivation to add this template is that it is default in HF if the model has no custom template: https://github.com/huggingface/transformers/blob/096f304695f7e7b169b031f7814352e900ad71c4/src/transformers/tokenization_utils_base.py#L1838